server/archive.api.js

```diff
@@ -0,0 +1,63 @@
+const HomebrewModel = require('./homebrew.model.js').model;
+const router = require('express').Router();
+const asyncHandler = require('express-async-handler');
+
+const archive = {
+	archiveApi : router,
+	/* Searches for matching title, also attempts to partial match */
+	findBrews  : async (req, res, next)=>{
+		try {
+			console.table(req.query);
+
+			const title = req.query.title || '';
+			const page = Math.max(parseInt(req.query.page) || 1, 1);
+			const minPageSize = 6; 
+			const pageSize = Math.max(parseInt(req.query.size) || 10, minPageSize);  
+			const skip = (page - 1) * pageSize;
+			
+			const titleQuery = {
+				title     : { $regex: decodeURIComponent(title), $options: 'i' },
+				$or: [],
+				published : true
+			};
+
+			if (req.query.legacy === 'true') {
+                titleQuery.$or.push({ renderer: 'legacy' });
+            };
+
+            if (req.query.v3 === 'true') {
+                titleQuery.$or.push({ renderer: 'V3' });
+            };
+
+			const projection = {
+				editId   : 0,
+				googleId : 0,
+				text     : 0,
+				textBin  : 0,
+			};
+			const brews = await HomebrewModel.find(titleQuery, projection)
+                .skip(skip)
+                .limit(pageSize)
+                .maxTimeMS(5000)
+                .exec();
+
+			const totalBrews = await HomebrewModel.countDocuments(titleQuery, projection).maxTimeMS(5000);
+            
+			const totalPages = Math.ceil(totalBrews / pageSize);
+            //console.log('Total brews: ', totalBrews);
+            //console.log('Total pages: ', totalPages);
+			return res.json({ brews, page, totalPages, totalBrews});
+		} catch (error) {
+			console.error(error);
+            if (error.response && error.response.status === 503) {
+                return res.status(503).json({ errorCode: '503', message: 'Service Unavailable' });
+            } else {
+                return res.status(500).json({ errorCode: '500', message: 'Internal Server Error' });
+            }
+		}
```

Comment:
OK, so here are my notes on improving title search...

Goals:

1. default to plain text search
  a) is faster than regex search. Because index.
  b) fewer errors from strings that are actually bogus regex patterns (e.g. unbalanced `[]` or `()`)
  c) avoid plain text searches (e.g. `[PF2e]` being handled as regex and getting unexpected results (e.g. `[PF2e]` as a regex search matches all titles that contain at least one of `P`, `F`, `2`, or `e` characters)

2. enable regex search by requiring the user to fence their regex pattern with `/`

3. for plain text title search, split the search string into words and do an AND search (so `WORD1 and WORD2` will match titles that contain both words, in any order, with any intervening words)

4. actually, split the plain text search string into operands, which are either space delimited words, or quoted strings. So we can do searches like `word1 word2 "quoted string"`.

5. detect if a plain text search operand has a `-` prefix, and interpret that as an exclude operand

So .. to start with #1 and #2...

```
  const title = req.query.title || '';
  const page = Math.max(parseInt(req.query.page) || 1, 1);
  const minPageSize = 6; 
  const pageSize = Math.max(parseInt(req.query.size) || 10, minPageSize);  
  const skip = (page - 1) * pageSize;

  const brewsQuery = {
    $or: [],
    published : true
  };

  if (req.query.v3 === 'true') { ... }
  if (req.query.legacy === 'true') { ... }
  
  if (title.startsWith('/') && title.endsWith('/')) {
    const titleConditionsArray = [
      {  "title": { $regex: title.slice(1, -1), $options: "i" } }
    ];  
  } else {
    // do plain text search
    const titleConditionsArray = buildTitleConditions (title); 
  }

  function buildTitleConditions(inputString) {
    return [ { "title": { $text: { $search: inputString, $caseSensitive: false } } } ];
  }
  
  const titleQuery = {
    $and: [
      brewsQuery,
      ...titleConditionsArray
    ]
  };

// ... now do the HomebrewModel.countDocuments() and HomebrewModel.find() and etc

  const projection = { ... };
  // ...

```


For #3, #4, #5 lets build out a fuller buildTitleConditions()

```
function buildTitleConditions(inputString) {
    const regex = /(-?"[^"]+"|-?\S+)/g;
    const matches = inputString.match(regex) || [];

    const finalQueries = [];

    for (const matchedString of matches) {
        const isExcluded = matchedString.startsWith('-');
        const baseString = isExcluded ? matchedString.slice(1) : matchedString;

        const unquotedMatch = baseString.startsWith('"') && baseString.endsWith('"')
            ? baseString.slice(1, -1)
            : baseString;

        const baseQuery = { "title": { $text: { $search: unquotedMatch, $caseSensitive: false } } };

        if (isExcluded) {
            finalQueries.push({ $not: baseQuery });
        } else {
            finalQueries.push(baseQuery);
        }
    }

    return finalQueries;
}

```

Reply:
Maybe i am wrong about this, but isn't titleConditionsArray defined inside an if, and tried to use down below, where it is 
not defined?

Reply:
This is implemented now, works fine in local, with some adjustments.